### PR TITLE
Follow the .clang-format style

### PR DESCRIPTION
**Pull request by ronen**
@steven-johnson I figured if there's a .clang-format style, may as well follow it.

In your most recent commit 143f76676d2e19f4c4b52ebfcbbeb6c2a48488f5 you wrote in the comment "This gets it a little closer to our de-facto style".  This PR would be making .clang-format be the de-jure style.

The first commit has clang-format -i run on everything.

Clang-format's alphabetizing of the #includes exposed a couple of files that had implicitly relied on external include order.  Fixed those in the second commit.

### CODE REVIEW

**Comment by steven-johnson** [comment]
I thought about doing this too, but decided it would definitely need consensus among all major code contributors; the style from clang-format is *close* but not identical. I'm guessing more people will chime in on whether they want this reformatting (in general) and if so if there are aspects of the formatting that need tweaking.

**Comment by abadams** [comment]
I like about 2/3 of the changes this makes and dislike the other third. I don't think it's worth the merge conflicts that it would create against the other branches.

**Comment by ronen** [comment]
> I like about 2/3 of the changes this makes and dislike the other third.

IMHO it's worth making it a goal to define the style by the `.clang-format` settings.   I think having the style be machine-conformable is very valuable.

It's super handy to be able to set up your editor to just style the code "correctly", rather than having to look up, learn, and manually conform to the official coding style of whatever project you happen to be working on at the moment.  

And it'd be similar handy when responding to PR's to be able just tell a contributor "please run `clang-format`" rather than having to educate them as to what the style is.  

Of course there are other aspects to style than just the text format, so this isn't a cure-all.  But still dealing with the text formatting is a big step.

So personally I'd suggest tweaking `.clang-format` til most people are happy with it then using that to define it the official standard.

(And I'll promise not to chime in on the `.clang-format` details discussion!  Personally, over the years I've done enough C++ projects needing to conform to different style standards that I've mostly gone numb and stopped caring what the standard is--as long as there is a standard.  IMHO most anything `clang-format` generates is a basically reasonable style that one could argue for/against... and can ultimately get used to.)  

> I don't think it's worth the merge conflicts that it would create against the other branches.

I started to write:

~I actually agree, did this mostly as an exercise...~

Then realized this is a case in point re benefits of a machine-conformable style:   Yes committing a change like this would cause merge conflicts with other branches.   But the conflicts would be very easily resolved--just run `clang-format` in the branch before merging!

More of an issue is that something in this seems to have broken the build... presumably some more order-dependence in the include files that didn't get picked up when I tested locally.  (Could set `SortIncludes: false` to avoid that but that would be cheating :)

File: apps/HelloHexagon/pipeline.cpp [hunk 1]
```diff
@@ -10,88 +10,84 @@ Expr blur5(Expr x0, Expr x1, Expr x2, Expr x3, Expr x4) {
...
-        blur(x, y, c) = blur5(blur_y(x - 2, y, c),
-                              blur_y(x - 1, y, c),
-                              blur_y(x,     y, c),
-                              blur_y(x + 1, y, c),
-                              blur_y(x + 2, y, c));
+        Func blur_y{ "blur_y" };
+        blur_y(x, y, c) =
+            blur5(input_bounded(x, y - 2, c), input_bounded(x, y - 1, c),
+                  input_bounded(x, y, c), input_bounded(x, y + 1, c),
+                  input_bounded(x, y + 2, c));
```
Comment (abadams) [inline apps/HelloHexagon/pipeline.cpp]: bad

File: apps/HelloHexagon/pipeline.cpp [hunk 2]
```diff
@@ -10,88 +10,84 @@ Expr blur5(Expr x0, Expr x1, Expr x2, Expr x3, Expr x4) {
...
                     .hexagon()
                     .prefetch(y, 2)
-                    .split(y, yo, y, 128).parallel(yo)
+                    .split(y, yo, y, 128)
+                    .parallel(yo)
                     .vectorize(x, vector_size * 2, TailStrategy::RoundUp);
-                blur_y
-                    .compute_at(blur, y)
-                    .vectorize(x, vector_size, TailStrategy::RoundUp);
+                blur_y.compute_at(blur, y).vectorize(x, vector_size,
```
Comment (abadams) [inline apps/HelloHexagon/pipeline.cpp]: very bad. Should have broken the line at the .

File: apps/HelloHexagon/pipeline.cpp [hunk 3]
```diff
@@ -10,88 +10,84 @@ Expr blur5(Expr x0, Expr x1, Expr x2, Expr x3, Expr x4) {
...
                 // Require scanlines of the input and output to be aligned.
-                input.dim(0).set_bounds(0, (input.dim(0).extent()/vector_size)*vector_size);
-                blur.dim(0).set_bounds(0, (blur.dim(0).extent()/vector_size)*vector_size);
+                input.dim(0).set_bounds(0, (input.dim(0).extent() / vector_size) * vector_size);
+                blur.dim(0).set_bounds(0, (blur.dim(0).extent() / vector_size) * vector_size);
 
                 for (int i = 1; i < 3; i++) {
-                    input.dim(i).set_stride((input.dim(i).stride()/vector_size)*vector_size);
-                    blur.dim(i).set_stride((blur.dim(i).stride()/vector_size)*vector_size);
+                    input.dim(i).set_stride((input.dim(i).stride() / vector_size) *
```
Comment (abadams) [inline apps/HelloHexagon/pipeline.cpp]: bad

File: apps/HelloHexagon/pipeline.cpp [hunk 4]
```diff
@@ -10,88 +10,84 @@ Expr blur5(Expr x0, Expr x1, Expr x2, Expr x3, Expr x4) {
...
                 }
             } else {
                 const int vector_size = natural_vector_size<uint8_t>();
 
-                Func(blur).compute_root()
-                .parallel(y, 16)
-                .vectorize(x, vector_size);
-                blur_y.compute_at(blur, y)
-                .vectorize(x, vector_size);
+                Func(blur).compute_root().parallel(y, 16).vectorize(x, vector_size);
```
Comment (abadams) [inline apps/HelloHexagon/pipeline.cpp]: bad. Should have preserved the line breaks but fixed the indentation

File: apps/HelloHexagon/process.cpp [hunk 5]
```diff
@@ -1,21 +1,23 @@
...
-#include "HalideRuntimeHexagonHost.h"
 #include "HalideBuffer.h"
+#include "HalideRuntimeHexagonHost.h"
 
-template <typename T>
+template<typename T>
 T clamp(T x, T min, T max) {
-    if (x < min) x = min;
-    if (x > max) x = max;
+    if (x < min)
```
Comment (abadams) [inline apps/HelloHexagon/process.cpp]: bad. multiline if statements require braces

File: apps/HelloMatlab/iir_blur.cpp [hunk 6]
```diff
@@ -45,12 +44,8 @@ Func blur_cols_transpose(Func input, Expr height, Expr alpha) {
     blur.compute_at(transpose, yo);
 
     // Vectorize computations within the strips.
-    blur.update(1)
-        .reorder(x, ry)
-        .vectorize(x);
-    blur.update(2)
-        .reorder(x, ry)
-        .vectorize(x);
+    blur.update(1).reorder(x, ry).vectorize(x);
```
Comment (abadams) [inline apps/HelloMatlab/iir_blur.cpp]: bad. Should not have removed line breaks

File: apps/HelloiOS/HelloiOS/reaction_diffusion_2_generator.cpp [hunk 7]
```diff
@@ -2,24 +2,20 @@
...
         if (get_target().has_gpu_feature()) {
-            Func(output)
-                .reorder(c, x, y)
-                .bound(c, 0, 3)
-                .vectorize(c)
-                .gpu_tile(x, y, 4, 4);
-            output
-                .dim(0).set_stride(3)
-                .dim(2).set_bounds(0, 3).set_stride(1);
+            Func(output).reorder(c, x, y).bound(c, 0, 3).vectorize(c).gpu_tile(x, y,
```
Comment (abadams) [inline apps/HelloiOS/HelloiOS/reaction_diffusion_2_generator.cpp]: very bad. Messed up the line breaks.

File: apps/HelloiOS/HelloiOS/reaction_diffusion_2_generator.cpp [hunk 8]
```diff
@@ -29,28 +25,25 @@ class ReactionDiffusion2Init : public Halide::Generator<ReactionDiffusion2Init>
...
-                           clamped(x, y, c) +
-                           clamped(x+1, y, c) +
-                           clamped(x+3, y, c));
-        blur_y(x, y, c) = (clamped(x, y-3, c) +
-                           clamped(x, y-1, c) +
-                           clamped(x, y, c) +
-                           clamped(x, y+1, c) +
-                           clamped(x, y+3, c));
-        blur(x, y, c) = (blur_x(x, y, c) + blur_y(x, y, c))/10;
+        blur_x(x, y, c) =
```
Comment (abadams) [inline apps/HelloiOS/HelloiOS/reaction_diffusion_2_generator.cpp]: bad

File: apps/HelloiOS/HelloiOS/reaction_diffusion_2_generator.cpp [hunk 9]
```diff
@@ -82,83 +75,62 @@ class ReactionDiffusion2Update : public Halide::Generator<ReactionDiffusion2Upda
...
                 .unroll(c)
                 .gpu_tile(clobber.x, clobber.y, 1, 1);
 
-            state
-                .dim(0).set_stride(3)
-                .dim(2).set_stride(1).set_extent(3);
-            new_state
-                .dim(0).set_stride(3)
-                .dim(2).set_stride(1).set_extent(3);
+            state.dim(0).set_stride(3).dim(2).set_stride(1).set_extent(3);
```
Comment (abadams) [inline apps/HelloiOS/HelloiOS/reaction_diffusion_2_generator.cpp]: bad

File: apps/HelloiOS/HelloiOS/reaction_diffusion_2_generator.cpp [hunk 10]
```diff
@@ -186,39 +161,30 @@ class ReactionDiffusion2Render : public Halide::Generator<ReactionDiffusion2Rend
...
 
         // Metal and CGImage require different pixel layouts.
         // Calculate both here and select() the right one;
         // we'll add specialize() paths in the schedule to
         // make this efficient.
-        Expr bgra = select(c == 0, cast<uint8_t>(B * 255), 
-                           c == 1, cast<uint8_t>(G * 255), 
-                           c == 2, cast<uint8_t>(R * 255),
-                           /*c==3*/cast<uint8_t>(A * 255));
+        Expr bgra = select(c == 0, cast<uint8_t>(B * 255), c == 1,
```
Comment (abadams) [inline apps/HelloiOS/HelloiOS/reaction_diffusion_2_generator.cpp]: bad

File: apps/HelloiOS/HelloiOS/reaction_diffusion_2_generator.cpp [hunk 11]
```diff
@@ -186,39 +161,30 @@ class ReactionDiffusion2Render : public Halide::Generator<ReactionDiffusion2Rend
...
+                           /*c==3*/ cast<uint8_t>(A * 255));
 
         render(x, y, c) = select(output_bgra == true, bgra, rgba);
     }
 
     void schedule() {
-        render
-            .dim(0).set_stride(4)
-            .dim(2).set_stride(1).set_bounds(0, 4);
+        render.dim(0).set_stride(4).dim(2).set_stride(1).set_bounds(0, 4);
```
Comment (abadams) [inline apps/HelloiOS/HelloiOS/reaction_diffusion_2_generator.cpp]: bad

File: apps/HelloiOS/HelloiOS/reaction_diffusion_2_generator.cpp [hunk 12]
```diff
@@ -186,39 +161,30 @@ class ReactionDiffusion2Render : public Halide::Generator<ReactionDiffusion2Rend
...
         if (get_target().has_gpu_feature()) {
-            state
-                .dim(0).set_stride(3)
-                .dim(2).set_stride(1).set_bounds(0, 3);
-            Func(render)
-                .reorder(c, x, y)
-                .unroll(c)
-                .gpu_tile(x, y, 32, 4);
+            state.dim(0).set_stride(3).dim(2).set_stride(1).set_bounds(0, 3);
```
Comment (abadams) [inline apps/HelloiOS/HelloiOS/reaction_diffusion_2_generator.cpp]: bad

File: apps/bilateral_grid/bilateral_grid_generator.cpp [hunk 13]
```diff
@@ -17,52 +17,52 @@ class BilateralGrid : public Halide::Generator<BilateralGrid> {
...
-                             blurz(x-1, y, z, c)*4 +
-                             blurz(x  , y, z, c)*6 +
-                             blurz(x+1, y, z, c)*4 +
-                             blurz(x+2, y, z, c));
-        blury(x, y, z, c) = (blurx(x, y-2, z, c) +
-                             blurx(x, y-1, z, c)*4 +
-                             blurx(x, y  , z, c)*6 +
-                             blurx(x, y+1, z, c)*4 +
-                             blurx(x, y+2, z, c));
+        blurz(x, y, z, c) =
```
Comment (abadams) [inline apps/bilateral_grid/bilateral_grid_generator.cpp]: bad

File: apps/bilateral_grid/bilateral_grid_generator.cpp [hunk 14]
```diff
@@ -82,27 +82,41 @@ class BilateralGrid : public Halide::Generator<BilateralGrid> {
             histogram.update().reorder(c, r.x, r.y, x, y).gpu_threads(x, y).unroll(c);
 
             // An alternative schedule for histogram that doesn't use shared memory:
-            // histogram.compute_root().reorder(c, z, x, y).gpu_tile(x, y, xi, yi, 8, 8);
-            // histogram.update().reorder(c, r.x, r.y, x, y).gpu_tile(x, y, xi, yi, 8, 8).unroll(c);
+            // histogram.compute_root().reorder(c, z, x, y).gpu_tile(x, y, xi, yi, 8,
```
Comment (abadams) [inline apps/bilateral_grid/bilateral_grid_generator.cpp]: very bad

File: apps/blur/test.cpp [hunk 15]
```diff
@@ -134,22 +131,24 @@ Buffer<uint16_t> blur_fast2(const Buffer<uint16_t> &in) {
...
+                __m128i *tmpPtr0 = tmp + ((y + 4) & 3) * vw;
+                __m128i *tmpPtr1 = tmp + ((y + 3) & 3) * vw;
+                __m128i *tmpPtr2 = tmp + ((y + 2) & 3) * vw;
                 for (int x = 0; x < vw; x++) {
                     // blur horizontally to produce next scanline of tmp
-                    __m128i val = _mm_load_si128((const __m128i *)(inPtr));
-                    val = _mm_add_epi16(val, _mm_loadu_si128((const __m128i *)(inPtr+1)));
-                    val = _mm_add_epi16(val, _mm_loadu_si128((const __m128i *)(inPtr+2)));
+                    __m128i val = _mm_load_si128((const __m128i *) (inPtr));
+                    val =
```
Comment (abadams) [inline apps/blur/test.cpp]: bad line break

File: apps/c_backend/run_cpp.cpp [hunk 16]
```diff
@@ -3,74 +3,73 @@
...
 #include "HalideBuffer.h"
-#include "pipeline_cpp_native.h"
 #include "pipeline_cpp_cpp.h"
+#include "pipeline_cpp_native.h"
 
 using namespace Halide::Runtime;
 
 extern "C" int an_extern_c_func(int a1, float a2) {
-    return (int)(a1 + a2);
+    return (int) (a1 + a2);
```
Comment (abadams) [inline apps/c_backend/run_cpp.cpp]: bad place to put a space

File: apps/cuda_mat_mul/mat_mul_generator.cpp [hunk 17]
```diff
@@ -36,20 +36,17 @@ class MatMul : public Halide::Generator<MatMul> {
...
-            .vectorize(B.in().args()[0]);
+        B.in().compute_at(prod, y).vectorize(B.in().args()[0]);
 
-        OutputImageParam bufs[] = {A, B, prod.output_buffer()};
+        OutputImageParam bufs[] = { A, B, prod.output_buffer() };
         for (auto &buf : bufs) {
-            buf.set_host_alignment(16)
-                .dim(0).set_bounds(0, size)
-                .dim(1).set_stride(size);
+            buf.set_host_alignment(16).dim(0).set_bounds(0, size).dim(1).set_stride(
```
Comment (abadams) [inline apps/cuda_mat_mul/mat_mul_generator.cpp]: very bad

File: apps/linear_algebra/benchmarks/halide_benchmarks.cpp [hunk 18]
```diff
@@ -84,103 +85,149 @@ struct BenchmarksBase {
...
-
-    L3Benchmark(gemm_transA, "d", halide_dgemm(true, false, alpha, A.raw_buffer(),
-                                                B.raw_buffer(), beta, C.raw_buffer()))
-
-    L3Benchmark(gemm_transB, "d", halide_dgemm(false, true, alpha, A.raw_buffer(),
-                                                B.raw_buffer(), beta, C.raw_buffer()))
-
-    L3Benchmark(gemm_transAB, "d", halide_dgemm(true, true, alpha, A.raw_buffer(),
-                                                B.raw_buffer(), beta, C.raw_buffer()))
+        L1Benchmark(scal, "d", halide_dscal(alpha, x.raw_buffer())) L1Benchmark(
```
Comment (abadams) [inline apps/linear_algebra/benchmarks/halide_benchmarks.cpp]: lolwut

**Comment by abadams** [comment]
If we can teach it a few more things are where we like line breaks, it might be workable. It just doesn't strike me as a great use of time and energy.

**Comment by steven-johnson** [comment]
> doesn't strike me as a great use of time and energy

I don't disagree -- main reason I updated the .clang-format file was because it was used for the first time in a long while for a recent change, and it was missing a few things that made it less-right than it could have been.

That said, this project has been mercifully free of formatting drama; when people make formatting mistakes, they get corrected, and minor inconsistencies are minor.

(If we could get clang-format to 90+% instead of 67% then yeah; not sure if current clang-format capabilities quite allow for our existing set of idiosyncrasies.)

**Comment by ronen** [comment]
@abadams I see what you're complaining about -- and actually it's my bad:  I shouldn't have included `apps/` in the reformatting.  I considered briefly whether to include `apps/` or just `src`, didn't put enough thought into it and decided what the heck just do everything.  And then I didn't look at the results in `apps/` only in `src/`

But `apps/` contains Halide code rather than ordinary C++ code.  And as became apparent when you looked at it (luckily `apps/` is first alphabetically!), reformatting of Halide code tends to fail badly for two reasons:  Halide source code is often laid out in 2D to convey the 2D structure of data.  And Halide code makes heavy use of method chaining, which clang-format doesn't have any good options for.

So I've just pushed a commit that reverts the reformatting in `apps/` and likewise `tutorial/`.

What's left behind is just the usual nitpicky little things like whitespace and indentation etc., that might be debatable (@abadams it seems you dislike `SpaceAfterCStyleCast: true`) but aren't catastrophic. And are what I had in mind in my argument in favor of all this.  And I think are likely to be in the 90+%  range.

If you're not sick of this topic, care to take another look?

Oh also I realize I was a bit glib when I wrote:

> But the conflicts would be very easily resolved--just run clang-format in the branch before merging!

There's a chicken-and-egg problem there: you need to merge to get the latest `.clang-format` file, but by then you've already gotten the merge conflicts.    One answer might be to make the approved `.clang-format` file easily available through a different channel, but that's a bit awkward.  Bears thinking about before doing a wholesale merge.

**Comment by abadams** [comment]
Sorry for the delay, I dropped everything that wasn't the halide_buffer_t PR for the last week or two.

I'm afraid I still think the changes in this PR, and more generally the idea of strictly enforcing use of clang-format, as a net negative change. The stuff it fixes is mostly very simple (e.g. spacing changes that make no significant difference to readability), but some of the stuff it changes gets pretty confusing (e.g. code alignment across lines, indentation of preprocessor macros). If we were to stick with it, I can anticipate fighting with it in the future to get it to accept how I want to lay something out. I already wage this battle regularly inside other codebases that have mechanically-enforced style. In larger projects it's probably a net win to synchronize style across hundreds or thousands of contributors, so I accept it elsewhere, but I think on balance our contributors are aligned enough on conventions, and already produce readable code. The other argument for mechanically-enforced style is to prevent arguments about style, but we haven't had any style arguments in the past that I can recall (though way back at the start of the project @jrk and I had an argument about how to indent '|' in ocaml that eventually required a coin flip to resolve.)

**Comment by jrk** [comment]
(…though the coin was clearly biased in favor my my approach because it was
the right answer. :)

**Comment by abadams** [comment]
You're a crazy person with dangerous ideas, and the default emacs ocaml
style agrees with me.

**Comment by steven-johnson** [comment]
> default emacs

You lost all credibility right there :-)

**Comment by abadams** [comment]
It's true. My .emacs file can fit on a single page :(

**Comment by ronen** [comment]
Emacs??  Really!?  (And nobody's made an Elisp target for Halide yet??  Where's that old-school emacs spirit?)

Anyway, getting back on topic...

> Sorry for the delay,

No prob.  This clearly isn't a critical issue.

> and more generally the idea of strictly enforcing use of clang-format, as a net negative change.

Fair enough.

> The stuff it fixes is mostly very simple (e.g. spacing changes that make no significant difference to readability),  ... I think on balance our contributors are aligned enough on conventions, and already produce readable code.

Speaking as a minor contributor, I had the recent experience  that I made a PR with what seemed to me to be readable code--and was requested to fix various small spacing changes etc that make no significant difference to readability :) in order to line up with the Halide coding style.    

I had a double reaction of:  *"Sure, whatever"*  because I don't really care--combined with a very slightly miffed *"Oh, but if there's a Halide coding style, how do I know what it is?  Where is it documented?  Why don't you have a `.clang-format` file for it?  Couldn't we save everybody this kind of back-and-forth?"*  And then I saw the `.clang-format `  file and felt sheepish that I hadn't used it in the first place.   And one thing led to another...

In any case, even if not rigorously enforcing clang-format everywhere, I do think it'd be good to have the `.clang-format` be approved as a baseline for the little white space things, so that a contributor can use it with confidence that the basic style will be in line with the rest of the codebase.  

I'm happy to close out this PR.   Unless you guys want to keep discussing emacs...

**Comment by steven-johnson** [comment]
> I do think it'd be good to have the .clang-format be approved as a baseline for the little white space things

+1. My vote would be:
-- Close out this PR
-- Augment the .clang-format and/or README files with notes about where and how to use it (mainly, good for new code from newcomers), with attention to the reasons we don't use it (some formatting issues we can't specify in clang-format at this time).
-- See how things progress; if revisions to clang-format eventually make it possible to specify an acceptable style, reconsider requiring it.

(If it's possible to take some of the benign changes separately, e.g. the minor spacing issues or the sorting of #includes, those could be nice to do as separate one-off cleanups IMHO)

**Comment by abadams** [comment]
+1 to that

**Comment by ronen** [comment]
Sounds good.  

> -- Close out this PR

Closing...

> -- Augment the .clang-format and/or README files with notes about where and how to use it (mainly, good for new code from newcomers), with attention to the reasons we don't use it (some formatting issues we can't specify in clang-format at this time).

If it's OK, I'll leave that in your(pl) capable hands, since you can explain/example better than I.

**Pull request closed without merging**